File: src-tauri/src/models/restore.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum RestoreOptionKind {
    OtaDowngrade,
    Powdersnow,
    Latest,
    BlobRestore,
    Tethered,
    CustomIpsw,
    DfuIpsw,
    SetNonce,
    IpswDownloader,
    MoreVersions,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RestoreOption {
    pub kind: RestoreOptionKind,
    pub title: String,
    pub description: String,
    pub target_version: Option<String>,
    pub requires_blobs: bool,
    pub requires_dfu: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RestoreOptionsResponse {
    pub product_type: Option<String>,
    pub processor_generation: Option<u8>,
    pub options: Vec<RestoreOption>,
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IpswDownloadRequest {
    pub url: String,
    pub output_dir: String,
    pub device_identifier: Option<String>,
    pub file_name: Option<String>,
    pub expected_sha1: Option<String>,
    pub download_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IpswDownloadResult {
    pub path: String,
    pub sha1: String,
    pub expected_sha1: Option<String>,
    pub sha1_matches: Option<bool>,
    pub download_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FirmwareListRequest {
    pub device_identifier: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FirmwareListEntry {
    pub version: String,
    pub build_id: String,
    pub url: String,
    pub sha1: Option<String>,
    pub size_bytes: Option<u64>,
    pub signed: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FirmwareListResult {
    pub device_identifier: String,
    pub fetched_at_unix: i64,
    pub cached: bool,
    pub firmwares: Vec<FirmwareListEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CheckIpswSigningRequest {
    pub device_identifier: String,
    pub build_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CheckIpswSigningResult {
    pub device_identifier: String,
    pub build_id: String,
    pub signed: bool,
    pub output: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CancelIpswDownloadRequest {
    pub download_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CancelIpswDownloadResult {
    pub download_id: String,
    pub cancelled: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IpswVerifyRequest {
    pub path: String,
    pub expected_sha1: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IpswVerifyResult {
    pub path: String,
    pub calculated_sha1: String,
    pub expected_sha1: Option<String>,
    pub matches: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum RestoreTool {
    IdeviceRestore,
    FutureRestore,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RestoreRunRequest {
    pub tool: RestoreTool,
    pub ipsw_path: String,
    pub shsh_path: Option<String>,
    pub erase: bool,
    pub update: bool,
    pub use_pwndfu: bool,
    pub skip_blob: bool,
    pub set_nonce: bool,
    pub no_baseband: bool,
    pub latest_sep: bool,
    pub latest_baseband: bool,
    pub dry_run: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RestoreCommandPreview {
    pub supported: bool,
    pub tool: RestoreTool,
    pub binary: String,
    pub args: Vec<String>,
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IpswPrepareRequest {
    pub ipsw_path: String,
    pub output_dir: String,
    pub device_identifier: Option<String>,
    pub shsh_path: Option<String>,
    pub device_ecid: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IpswPrepareResult {
    pub output_path: String,
}

File: src/lib/api/restore.ts
import { invoke } from '@tauri-apps/api/core';
import { listen, type UnlistenFn } from '@tauri-apps/api/event';
import type { DeviceInfo } from '$lib/stores/deviceStore.svelte';

export type RestoreOptionKind =
  | 'otaDowngrade'
  | 'powdersnow'
  | 'latest'
  | 'blobRestore'
  | 'tethered'
  | 'customIpsw'
  | 'dfuIpsw'
  | 'setNonce'
  | 'ipswDownloader'
  | 'moreVersions';

export interface RestoreOption {
  kind: RestoreOptionKind;
  title: string;
  description: string;
  targetVersion: string | null;
  requiresBlobs: boolean;
  requiresDfu: boolean;
}

export interface RestoreOptionsResponse {
  productType: string | null;
  processorGeneration: number | null;
  options: RestoreOption[];
  warnings: string[];
}

export interface IpswDownloadRequest {
  url: string;
  outputDir: string;
  deviceIdentifier?: string | null;
  fileName: string | null;
  expectedSha1: string | null;
  downloadId: string | null;
}

export interface IpswDownloadResult {
  path: string;
  sha1: string;
  expectedSha1: string | null;
  sha1Matches: boolean | null;
  downloadId: string;
}

export interface FirmwareListRequest {
  deviceIdentifier: string;
}

export interface FirmwareListEntry {
  version: string;
  buildId: string;
  url: string;
  sha1: string | null;
  sizeBytes: number | null;
  signed: boolean | null;
}

export interface FirmwareListResult {
  deviceIdentifier: string;
  fetchedAtUnix: number;
  cached: boolean;
  firmwares: FirmwareListEntry[];
}

export interface CheckIpswSigningRequest {
  deviceIdentifier: string;
  buildId: string;
}

export interface CheckIpswSigningResult {
  deviceIdentifier: string;
  buildId: string;
  signed: boolean;
  output: string;
}

export interface CancelIpswDownloadRequest {
  downloadId: string;
}

export interface CancelIpswDownloadResult {
  downloadId: string;
  cancelled: boolean;
}

export interface IpswDownloadProgressEvent {
  downloadId: string;
  percent: number | null;
  downloadedBytes: number | null;
  totalBytes: number | null;
  speedBps: number | null;
  etaSeconds: number | null;
}

export interface IpswVerifyRequest {
  path: string;
  expectedSha1: string | null;
}

export interface IpswVerifyResult {
  path: string;
  calculatedSha1: string;
  expectedSha1: string | null;
  matches: boolean | null;
}

export type RestoreTool = 'ideviceRestore' | 'futureRestore';

export interface RestoreRunRequest {
  tool: RestoreTool;
  ipswPath: string;
  shshPath: string | null;
  erase: boolean;
  update: boolean;
  usePwndfu: boolean;
  skipBlob: boolean;
  setNonce: boolean;
  noBaseband: boolean;
  latestSep: boolean;
  latestBaseband: boolean;
  dryRun: boolean;
}

export interface RestoreCommandPreview {
  supported: boolean;
  tool: RestoreTool;
  binary: string;
  args: string[];
  warnings: string[];
}

export interface IpswPrepareRequest {
  ipswPath: string;
  outputDir: string;
  deviceIdentifier?: string | null;
  shshPath: string | null;
  deviceEcid: string | null;
}

export interface IpswPrepareResult {
  outputPath: string;
}

export function getRestoreOptions(device: DeviceInfo): Promise<RestoreOptionsResponse> {
  return invoke<RestoreOptionsResponse>('get_restore_options', { device });
}

export function downloadIpsw(request: IpswDownloadRequest): Promise<IpswDownloadResult> {
  return invoke<IpswDownloadResult>('download_ipsw', { request });
}

export function listFirmwares(request: FirmwareListRequest): Promise<FirmwareListResult> {
  return invoke<FirmwareListResult>('list_firmwares', { request });
}

export interface ExistingIpswEntry {
  path: string;
  fileName: string;
  sizeBytes: number;
  deviceIdentifier: string | null;
}

export interface ListExistingIpswsRequest {
  deviceIdentifier?: string | null;
}

export interface ListExistingIpswsResult {
  ipsws: ExistingIpswEntry[];
}

export function listExistingIpsws(
  request: ListExistingIpswsRequest,
): Promise<ListExistingIpswsResult> {
  return invoke<ListExistingIpswsResult>('list_existing_ipsws', { request });
}

export function checkIpswSigning(
  request: CheckIpswSigningRequest,
): Promise<CheckIpswSigningResult> {
  return invoke<CheckIpswSigningResult>('check_ipsw_signing', { request });
}

export function cancelIpswDownload(
  request: CancelIpswDownloadRequest,
): Promise<CancelIpswDownloadResult> {
  return invoke<CancelIpswDownloadResult>('cancel_ipsw_download', { request });
}

export function onIpswDownloadProgress(
  handler: (event: IpswDownloadProgressEvent) => void,
): Promise<UnlistenFn> {
  return listen<IpswDownloadProgressEvent>('ipsw-download-progress', (e) => handler(e.payload));
}

export function verifyIpsw(request: IpswVerifyRequest): Promise<IpswVerifyResult> {
  return invoke<IpswVerifyResult>('verify_ipsw', { request });
}

export function prepareIpsw(request: IpswPrepareRequest): Promise<IpswPrepareResult> {
  return invoke<IpswPrepareResult>('prepare_ipsw', { request });
}

export function previewRestoreCommand(request: RestoreRunRequest): Promise<RestoreCommandPreview> {
  return invoke<RestoreCommandPreview>('preview_restore_command', { request });
}

export function startRestore(request: RestoreRunRequest): Promise<RestoreCommandPreview> {
  return invoke<RestoreCommandPreview>('start_restore', { request });
}
